### PR TITLE
Core issuance flow

### PR DESCRIPTION
**Pull request by asoong**
Core issuance flow. Some interesting diffs: decided not to call deposit in issuance in order to save gas on the extra `incrementVaultOwner` call for the deposited amounts. Just subtract the amount of the component that was already in there.

Changed `STANDARD_INITIAL_TOKENS` constant to `DEPLOYED_TOKEN_QUANTITY`

### CODE REVIEW

File: test/logs/log_utils.ts [hunk 1]
```diff
@@ -45,7 +45,19 @@ export function formatLogEntry(logs: ABIDecoder.DecodedLog): Log {
...
   _.each(events, (event) => {
     const { name, type, value } = event;
-    args[name] = (/^(uint)\d*/.test(type)) ? new BigNumber(value.toString()) : value;
+
+    var argValue: any = value;
+    switch (true) {
+      case (/^(uint)\d*\[\]/.test(type)): {
+          break;
+       }
+       case (/^(uint)\d*/.test(type)): {
```
Comment (asoong) [inline test/logs/log_utils.ts]: Fixed indentation.

File: contracts/core/Core.sol [hunk 2]
```diff
@@ -77,14 +94,23 @@ contract Core is
...
+    modifier isValidFactory(address _factoryAddress) {
         require(
-            isValidFactory[_factoryAddress],
+            validFactories[_factoryAddress],
             INVALID_FACTORY
         );
         _;
     }
 
+    // Verify set was created by core and is enabled for issuance
```
Comment (justinkchen) [inline contracts/core/Core.sol]: don't mention issuance. i use the same for redeem

File: contracts/core/Core.sol [hunk 3]
```diff
@@ -69,6 +77,15 @@ contract Core is
 
     /* ============ Modifiers ============ */
 
+    // Validate quantity is multiple of natural unit
+    modifier isValidQuantity(uint _quantity, address _setToken) {
```
Comment (justinkchen) [inline contracts/core/Core.sol]: lets mention natural unit in the modifier, `isValidQuantity` of what? we have another "quantity verifier" right below with nonZero check. maybe `isNaturalUnitMultiple`?

File: contracts/core/Core.sol [hunk 4]
```diff
@@ -166,11 +192,85 @@ contract Core is
...
-        isValidFactory[_factoryAddress] = false;
+        validFactories[_factoryAddress] = false;
     }
 
     /* ============ Public Functions ============ */
 
+    /**
+     * Issue
+     *
+     * @param  _setAddress    Array of the addresses of the ERC20 tokens
```
Comment (justinkchen) [inline contracts/core/Core.sol]: these descriptions seem off?

File: contracts/core/Vault.sol [hunk 5]
```diff
@@ -154,6 +154,7 @@ contract Vault is
     }
 
     /* ============ Getter Functions ============ */
+
```
Comment (justinkchen) [inline contracts/core/Vault.sol]: remove

Comment (justinkchen) [reply contracts/core/Vault.sol]: nvm this space is on purpose

File: contracts/core/Core.sol [hunk 6]
```diff
@@ -166,11 +192,85 @@ contract Core is
...
+            if (amountToDecrement > 0) {
+                IVault(vaultAddress).decrementTokenOwner(
+                    msg.sender,
+                    component,
+                    amountToDecrement
+                );
+            }
+
+            // Increment the vault balance of the set token for the component
+            IVault(vaultAddress).incrementTokenOwner(
```
[the remark this answers is not in the document]
Comment (asoong) [reply contracts/core/Core.sol]: this is to tell us the balance of the components in the vault that belong to the set
the set token itself keeps the balance of each user

File: contracts/core/Core.sol (hunk 4 above)
Comment (asoong) [reply contracts/core/Core.sol]: +1

File: contracts/core/Core.sol (hunk 3 above)
Comment (asoong) [reply contracts/core/Core.sol]: +1

File: contracts/core/Core.sol (hunk 2 above)
Comment (asoong) [reply contracts/core/Core.sol]: +1

File: test/logs/log_utils.ts [hunk 7]
```diff
@@ -45,7 +45,19 @@ export function formatLogEntry(logs: ABIDecoder.DecodedLog): Log {
   // Loop through each event and add to args
   _.each(events, (event) => {
     const { name, type, value } = event;
-    args[name] = (/^(uint)\d*/.test(type)) ? new BigNumber(value.toString()) : value;
+
+    var argValue: any = value;
+    switch (true) {
+      case (/^(uint)\d*\[\]/.test(type)): {
+        break;
```
Comment (felix2feng) [inline test/logs/log_utils.ts]: Does this work? Don't we want to set the `argValue` to be an array of BigNumbers if this is the case?

Comment (asoong) [reply test/logs/log_utils.ts]: Yup, it worked for me

File: test/core.spec.ts [hunk 8]
```diff
@@ -678,7 +683,209 @@ contract("Core", (accounts) => {
...
+    let setToken: SetTokenContract;
+
+    beforeEach(async () => {
+      await deployCoreAndInitializeDependencies();
+
+      components = await coreWrapper.deployTokensAsync(2, ownerAccount);
+      await coreWrapper.approveTransfersAsync(components, transferProxy.address);
+
+      const componentAddresses = _.map(components, (token) => token.address);
+      componentUnits = _.map(components, () => ether(4)); // Multiple of of naturalUnit
```
[the remark this answers is not in the document]
Comment (asoong) [reply test/core.spec.ts]: +1

**Pull request merged**